File: hanoi.cpp
/* See LICENSE file for copyright and license details. */

#include <iostream>

/* move n-th disk from `from` to `to` */
void move(long n, char from, char to, char via)
{
	if (n > 0) {
		move(n - 1, from, via, to);
		std::cout << n << ": " << from << " → " << to << '\n';
		move(n - 1, via, to, from);
	}
}

int main(void)
{
	long n;
	std::cout << "Number of disks: ";
	std::cin >> n;
	if (std::cin.fail() || n < 1) {
		std::cerr << "Error: invalid input\n";
		return EXIT_FAILURE;
	}

	move(n, 'a', 'c', 'b');
	return EXIT_SUCCESS;
}

File: hanoi.py
# See LICENSE file for copyright and license details.

import sys

# move n-th disk from `frm` to `to`
def move(n, frm, to, via):
    if n > 0:
        move(n - 1, frm, via, to)
        print('{0}: {1} → {2}'.format(n, frm, to))
        move(n - 1, via, to, frm)

inp = input('Number of disks: ')
if not inp.isdigit() or inp == '0':
    sys.exit('Error: invalid input')
n = int(inp)

move(n, 'a', 'c', 'b');
